File: Motocitizen/src/motocitizen/activity/MainScreenActivity.java
package motocitizen.activity;

import android.Manifest;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.listener.PermissionGrantedResponse;
import com.karumi.dexter.listener.single.BasePermissionListener;

import motocitizen.MyApp;
import motocitizen.content.Content;
import motocitizen.content.accident.Accident;
import motocitizen.geolocation.MyLocationManager;
import motocitizen.main.R;
import motocitizen.maps.MyMapManager;
import motocitizen.maps.google.MyGoogleMapManager;
import motocitizen.router.Router;
import motocitizen.rows.RowFactory;
import motocitizen.user.User;
import motocitizen.utils.BounceScrollView;
import motocitizen.utils.ChangeLog;
import motocitizen.utils.Preferences;
import motocitizen.utils.Utils;

public class MainScreenActivity extends AppCompatActivity implements MyFragmentInterface {
    private static final byte LIST = 0;
    private static final byte MAP  = 1;

    private ViewGroup   mapContainer;
    private ImageButton createAccButton;
    private ImageButton toAccListButton;
    private ImageButton toMapButton;
    private View        accListView;
    private ProgressBar progressBar;
    private ViewGroup   listContent;

    private MenuItem     refreshItem;
    private MyMapManager map;
    private boolean inTransaction = false;
    private byte    currentScreen = LIST;
    private static ActionBar actionBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_screen_activity);
        actionBar = getSupportActionBar();
    }

    @Override
    protected void onResume() {
        super.onResume();
        MyLocationManager.getInstance().wakeup(this);

        if (Preferences.Companion.getInstance(this).getNewVersion()) {
            AlertDialog changeLogDlg = ChangeLog.INSTANCE.getDialog(this);
            changeLogDlg.show();
            Preferences.Companion.getInstance(this).setNewVersion(false);
        }
        PackageManager pm = this.getPackageManager();
        if (!pm.hasSystemFeature(PackageManager.FEATURE_TELEPHONY)) {
            this.findViewById(R.id.dial_button).setEnabled(false);
        }

        if (accListView == null) accListView = this.findViewById(R.id.acc_list);
        if (progressBar == null)
            progressBar = (ProgressBar) this.findViewById(R.id.progressBar);
        if (mapContainer == null)
            mapContainer = (ViewGroup) this.findViewById(R.id.google_map);
        if (createAccButton == null)
            createAccButton = (ImageButton) this.findViewById(R.id.add_point_button);
        if (toAccListButton == null)
            toAccListButton = (ImageButton) this.findViewById(R.id.list_button);
        if (toMapButton == null)
            toMapButton = (ImageButton) this.findViewById(R.id.map_button);

        createAccButton.setOnClickListener(v -> Router.INSTANCE.goTo(this, Router.Target.CREATE));
        toAccListButton.setOnClickListener(v -> setScreen(LIST));
        toMapButton.setOnClickListener(v -> setScreen(MAP));
        this.findViewById(R.id.dial_button).setOnClickListener(v -> Router.INSTANCE.dial(this, getString(R.string.phone)));
        ((BounceScrollView) this.findViewById(R.id.accListRefresh)).setOverScrollListener(this::getAccidents);
        listContent = (ViewGroup) this.findViewById(R.id.accListContent);

        if (map == null) map = new MyGoogleMapManager(this);
        Dexter.withActivity(this)
              .withPermission(Manifest.permission.ACCESS_FINE_LOCATION)
              .withListener(new BasePermissionListener() {
                  @Override
                  public void onPermissionGranted(PermissionGrantedResponse response) {
                      map.enableLocation();
                  }
              }).check();

        setPermissions();
        setScreen(currentScreen);
        redraw();
        getAccidents();
    }

    @Override
    protected void onPause() {
        super.onPause();
        MyLocationManager.getInstance().sleep(this);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (intent.hasExtra("toMap")) {
            //mainFragment.toMap(intent.getExtras().getInt("toMap", 0));
            toMap(intent.getExtras().getInt("toMap", 0));
            intent.removeExtra("toMap");
        }
        if (intent.hasExtra("toDetails")) {
            intent.removeExtra("toDetails");
        }
        setIntent(intent);
    }

    public static void updateStatusBar(String address) {
        String subTitle = "";
        //Делим примерно пополам, учитывая пробел или запятую
        int commaPos = address.lastIndexOf(",", address.length() / 2);
        int spacePos = address.lastIndexOf(" ", address.length() / 2);

        if (commaPos != -1 || spacePos != -1) {
            subTitle = address.substring(Math.max(commaPos, spacePos) + 1);
            address = address.substring(0, Math.max(commaPos, spacePos));
        }

        actionBar.setTitle(address);
        if (!subTitle.isEmpty()) actionBar.setSubtitle(subTitle);
    }

    @Override
    public void setPermissions() {
        createAccButton.setVisibility(User.getInstance(this).isStandard() ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    public void redraw() {
        listContent.removeAllViews();

        //TODO YesterdayRow ???
        //TODO Нет событий

        for (Accident accident : Content.INSTANCE.getListReversed()) {
            if (accident.isInvisible(this)) continue;
            listContent.addView(RowFactory.INSTANCE.make(this, accident));
        }
        map.placeAccidents(this);
    }

    private void getAccidents() {
        if (inTransaction) return;
        if (MyApp.isOnline(this)) {
            startRefreshAnimation();
            Content.INSTANCE.requestUpdate(result -> {
                this.runOnUiThread(() -> {
                    stopRefreshAnimation();
                    redraw();
                });
            });
        } else {
            Toast.makeText(this, getString(R.string.inet_not_available), Toast.LENGTH_LONG).show();
        }
    }

    private void setRefreshAnimation(boolean status) {
        progressBar.setVisibility(status ? View.VISIBLE : View.INVISIBLE);
        inTransaction = status;
        //TODO костыль
        if (refreshItem != null) refreshItem.setVisible(!status);
    }

    private void stopRefreshAnimation() {
        setRefreshAnimation(false);
    }

    private void startRefreshAnimation() {
        setRefreshAnimation(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.small_settings_menu, menu);
        refreshItem = menu.findItem(R.id.action_refresh);
        if (inTransaction) refreshItem.setVisible(false);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.small_menu_refresh:
                getAccidents();
                return true;
            case R.id.small_menu_settings:
                Router.INSTANCE.goTo(this, Router.Target.SETTINGS);
                return true;
            case R.id.small_menu_about:
                Router.INSTANCE.goTo(this, Router.Target.ABOUT);
                return true;
            case R.id.small_menu_exit:
                Router.INSTANCE.exit(this);
                return true;
            case R.id.action_refresh:
                getAccidents();
                return true;
            case R.id.do_not_disturb:
                item.setIcon(Preferences.Companion.getInstance(this).getDoNotDisturb() ? R.drawable.ic_lock_ringer_on_alpha : R.drawable.ic_lock_ringer_off_alpha);
                Preferences.Companion.getInstance(this).setDoNotDisturb(!Preferences.Companion.getInstance(this).getDoNotDisturb());
                return true;
        }
        return false;
    }

    private void setScreen(byte target) {
        currentScreen = target;
        toAccListButton.setAlpha(target == LIST ? 1f : 0.3f);
        toMapButton.setAlpha(target == MAP ? 1f : 0.3f);
        accListView.animate().translationX(target == LIST ? 0 : -Utils.getWidth(this) * 2);
        mapContainer.animate().translationX(target == MAP ? 0 : Utils.getWidth(this) * 2);
    }

    public void toMap(int id) {
        setScreen(MAP);
        map.jumpToPoint(Content.INSTANCE.getAccidents().get(id).getLocation());
    }
}

File: Motocitizen/src/motocitizen/ui/activity/MainScreenActivity.kt
package motocitizen.ui.activity

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintSet.WRAP_CONTENT
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import motocitizen.MyApp
import motocitizen.content.Content
import motocitizen.content.accident.Accident
import motocitizen.datasources.preferences.Preferences
import motocitizen.geo.geolocation.MyLocationManager
import motocitizen.geo.maps.MainMapManager
import motocitizen.main.R
import motocitizen.permissions.Permissions
import motocitizen.subscribe.SubscribeManager
import motocitizen.ui.Screens
import motocitizen.ui.changelog.ChangeLog
import motocitizen.ui.rows.AccidentRow
import motocitizen.ui.views.BounceScrollView
import motocitizen.user.User
import motocitizen.utils.BackgroundScope
import motocitizen.utils.bindView
import motocitizen.utils.carryOver
import motocitizen.utils.displayWidth
import motocitizen.utils.goTo
import motocitizen.utils.hide
import motocitizen.utils.makeDial
import motocitizen.utils.margins
import motocitizen.utils.padToolbars
import motocitizen.utils.show

class MainScreenActivity : AppCompatActivity() {
    companion object {
        private const val LIST: Byte = 0
        private const val MAP: Byte = 1
        private const val SUBSCRIBE_TAG = "mainScreen"
    }

    private val listenersScope = BackgroundScope.Default()

    private val mapContainer: ViewGroup by bindView(R.id.google_map)
    private val createAccButton: ImageButton by bindView(R.id.add_point_button)
    private val toAccListButton: ImageButton by bindView(R.id.list_button)
    private val toMapButton: ImageButton by bindView(R.id.map_button)
    private val accListView: View by bindView(R.id.acc_list)
    private val progressBar: ProgressBar by bindView(R.id.progressBar)
    private val listContent: ViewGroup by bindView(R.id.accListContent)
    private val dialButton: ImageButton by bindView(R.id.dial_button)

    private val bounceScrollView: BounceScrollView by bindView(R.id.accListRefresh)

    private var refreshItem: MenuItem? = null
    private lateinit var map: MainMapManager
    private var currentScreen = LIST

    private var transaction: Boolean = false
        set(value) {
            field = value
            refreshItem?.isVisible = !value
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_screen_activity)
        map = MainMapManager(this)
        ChangeLog.show(this)
        padToolbars(accListView)
        padToolbars(mapContainer)
    }

    override fun onResume() {
        super.onResume()

        runBlocking {
            showCurrentFrame()
            setUpFeaturesAccessibility()
            val redraw = async { redraw() }
            setUpListeners()
            subscribe()
            launch {
                MyLocationManager.wakeup(this@MainScreenActivity)
            }
            redraw.await()
            requestAccidents()
        }
    }

    private fun showCurrentFrame() {
        setFrame(currentScreen)
    }

    //todo refactor
    private fun setFrame(target: Byte) {
        currentScreen = target
        toAccListButton.alpha = if (target == LIST) 1f else 0.3f
        toMapButton.alpha = if (target == MAP) 1f else 0.3f
        accListView.animate()
            .translationX((if (target == LIST) 0 else -displayWidth() * 2).toFloat())
        mapContainer.animate()
            .translationX((if (target == MAP) 0 else displayWidth() * 2).toFloat())
    }

    private fun redraw() {
        val newList = Content.getVisibleReversed()
            .map { accident ->
                AccidentRow.make(this@MainScreenActivity, accident).apply {
                    layoutParams = LinearLayout.LayoutParams(MATCH_PARENT, WRAP_CONTENT).margins(preset.margins)
                    setOnClickListener { toMap(accident) }
                }
            }

        runOnUiThread {
            listContent.removeAllViews()
            newList.forEach(listContent::addView)
            map.update()
        }
    }

    private fun setUpFeaturesAccessibility() = runOnUiThread {
        createAccButton.apply { if (!User.isReadOnly()) show() else hide() }
        dialButton.isEnabled = packageManager.hasSystemFeature(PackageManager.FEATURE_TELEPHONY)
    }

    private fun setUpListeners() = listenersScope.launch {
        createAccButton.setOnClickListener { goTo(Screens.CREATE) }
        toAccListButton.setOnClickListener { setFrame(LIST) }
        toMapButton.setOnClickListener { setFrame(MAP) }
        dialButton.setOnClickListener { makeDial(getString(R.string.phone)) }
        bounceScrollView.setOverScrollListener { lifecycleScope.launch { requestAccidents() } }
    }

    private fun subscribe() = listenersScope.launch {
        SubscribeManager.subscribe(
            SubscribeManager.Event.LOCATION_UPDATED,
            SUBSCRIBE_TAG
        ) { updateStatusBar() }
        SubscribeManager.subscribe(
            SubscribeManager.Event.ACCIDENTS_UPDATED,
            SUBSCRIBE_TAG
        ) { listenersScope.launch { redraw() } }
    }

    private suspend fun requestAccidents() {
        when {
            transaction -> return
            !MyApp.isOnline(this) -> Toast.makeText(
                this,
                getString(R.string.inet_not_available),
                Toast.LENGTH_LONG
            ).show()

            else -> {
                transaction = true
                progressBar.show()
                Content.requestUpdate { updateCompleteCallback() }
            }
        }
    }

    private fun updateCompleteCallback() = runOnUiThread {
        redraw()
        progressBar.hide()
        transaction = false
    }

    override fun onPause() {
        super.onPause()
        SubscribeManager.unSubscribeAll(SUBSCRIBE_TAG)
        Permissions.requestLocation(this) { MyLocationManager.sleep() }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        intent.apply {
            when {
                hasExtra("toMap") -> toMap(intent.extras?.getInt("toMap", 0) ?: 0)
            }
            removeExtra("toMap")
        }
        setIntent(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.small_settings_menu, menu)
        refreshItem = menu.findItem(R.id.action_refresh)
        if (transaction) refreshItem?.isVisible = false

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.small_menu_settings -> goTo(Screens.SETTINGS)
            R.id.small_menu_about -> goTo(Screens.ABOUT)
            R.id.action_refresh -> lifecycleScope.launch { requestAccidents() }
            R.id.do_not_disturb -> flipDoNotDisturb(item)
            else -> return false
        }
        return true
    }

    private fun flipDoNotDisturb(item: MenuItem) {
        item.setIcon(if (Preferences.doNotDisturb) R.drawable.ic_lock_ringer_on_alpha else R.drawable.ic_lock_ringer_off_alpha)
        Preferences.doNotDisturb = !Preferences.doNotDisturb
    }

    private fun toMap(id: Int) {
        toMap(Content[id] ?: return)
    }

    private fun toMap(accident: Accident) {
        setFrame(MAP)
        map.centerOnAccident(accident)
    }

    private fun updateStatusBar() {
        val (part1, part2) = MyLocationManager.getAddress().carryOver()

        actionBar?.apply {
            title = part1
            subtitle = part2
        }
    }
}
